### ISSUE TITLE
Store more information in history

### ISSUE THREAD

**Issue by aartaka** [atlas-engineer/nyxt#1848]
[https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/history](The WebExtensions history API) requires the browser to store all the visits to a URL, with the respective timestamp and type of navigation. This hints at the fact that we can store more information in history than we currently do, and that we'll eventually need to do it.

The information to store (see the link above):
- [ ] Every individual visit, with time and navigation type.
- [ ] Downloads.
- For history items ([history.HistoryItem spec](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/history/HistoryItem)):
  - [ ] Number of times the URL was explicitly typed in `set-url` prompt.
  - [ ] Scroll position (broken ATM).
  - [ ] [Zoom level.](https://github.com/atlas-engineer/nyxt/pull/3036#discussion_r1236602602)
- Navigation types (taken from the [WebExtensions history.TransitionType specs](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/history/TransitionType)):
  - [ ] Link click.
  - [ ] Direct `set-url` prompt typing.
  - [ ] Bookmark.
  - [ ] iframes (loaded by the page or the user).
  - [ ] Command line passing.
  - [ ] Form submission.
  - [ ] Page reloading.
  - [ ] Keyword search (what? something Firefox-specific here)

More so, this spec requires history items to be easily deletable. This can be problematic with our immutable history.

There are obviously more things we can add to our history (see [Tausher&Greenberg](https://www.sciencedirect.com/science/article/pii/S1071581997901257) for an exhaustive list of things like semantic namespaces for history and other exciting stuff), but those changes are what every other major browser has and provides to its extensions. Thus, the changes listed above are first things to improve in our history.

**Comment by Ambrevar** [comment]
I'm confused by

> There are obviously more things we can add to our history (see Tausher&Greenberg for an exhaustive list of things like semantic namespaces for history and other exciting stuff), but those changes are what every other major browser has and provides to its extensions. 

Are you saying that major browsers have semantic namespaces?

**Comment by Ambrevar** [comment]
Off the top of my head, I remember that we can store arbitrary data to our history entries, so most of this should be easy, no?

As for entry deletion, do you know an example where it would be really needed?

**Comment by aartaka** [comment]
> I'm confused by
> 
> > There are obviously more things we can add to our history (see Tausher&Greenberg for an exhaustive list of things like semantic namespaces for history and other exciting stuff), but those changes are what every other major browser has and provides to its extensions.
> 
> Are you saying that major browsers have semantic namespaces?

Sorry, I've been unclear. What I mean is that the linked paper has some insights for features that even major browsers don't have. The list above the reference to the paper, though, is standardized and all the Chromium- and Gecko-based browsers should have it.

**Comment by aartaka** [comment]
> Off the top of my head, I remember that we can store arbitrary data to our history entries, so most of this should be easy, no?

Yes. The only thing that puzzles me is how to store visits, as right now this is only represented as `implicit-visits` counter, while it should be a collection of visit objects with their own properties. How and where to implement it and where to add it are relatively non-trivial questions.

> As for entry deletion, do you know an example where it would be really needed?

Extensions that rely on `history` WebExtensions API can use it for whatever reason they want it. The mere fact of this API existence is problematic there -- The Law of Implicit APIs ensures that if it exists then it's used by someone. While we can diverge from the standard and error out on history deletion when it's not allowed, it's not nice to existing extensions.

**Comment by aadcg** [comment]
No longer important, due to #3598.
